tools/codegen-crds/README.md

```diff
@@ -0,0 +1,81 @@
+# Codegen-crds
+
+This tool simplifies the process of generating CRD schemas across a large number of API groups and versions.
+It's aim is to encode specific knowledge about the schemapatch generators we use and the yaml patch methods we include
+in this repo and handle the complexity around running those.
+
+## Usage
+
+The tool can be compiled in the normal way using either `go build` or `make codegen-crds` from the root of the tools
+module.
+When using make the tool will output to `$(TOOLS_MODULE)/_output/bin/$(GOOS)/$(GOARCH)/bin/codeged-crds`.
+
+The function has the following arguments:
+- `--api-group-versions` - A comma separated list of group versions to generate, e.g. apps/v1,machine/v1,machine/v1beta1.
+  Each group version will be loaded and the schema patch will be generated into the pre-existing CRD.
+  Note, if the CRD YAML file does not already exist, it will not be generated for you.
+- `--base-dir` - The path to the root of the API folders, this should be the parent of the group versions, e.g. if your
```

Comment:
if we're re-imagining this, why basedir+api-group-version instead of golang packages to the api group versions?  Honest question about why we would choose A over B.

I see a few dimensions
1. which packages do you want to generate for
2. how do individual versions generate
3. which versions are you going to generate into a single output folder

I think 1 and 3 are related and may be composed independent of 2.  

An example of 3 would be something like the openapi in this package or the group-based separation for clients in openshift/client-go.

Reply:
> instead of golang packages

If we use the full golang packages this makes it difficult for a lot of folks who have different environments set up, eg not having GOPATH. The current approach is making everything work based on absolute or relative paths which should work wherever you are, where the golang package approach which is adopted in a lot of the generators today has made it difficult to allow flexibility in folks environments, and I know some RH'ers have been unpicking that upstream to make things work outside of GOPATH going forward.

I agree with the dimensions you've outlined btw, though package vs folder could be argued depending on your view

Reply:
> I'm willing to let this develop in either direction (package or folder), but choosing some consistent way to locate has benefits

I'll look into the various generators upstream and determine what it is that they expect to see, if we go down the package route, I do worry that relies on a lot of "go" internals to look up the packages and requires us to parse the entire "base dir", which could be expensive.

I'll read into this and get back to you, but yes, something consistent for all the generators and tools would be helpful to the end users here

> actual groupname (with the dots) and versions

You mean a user would provider, `machine.openshift.io/v1` as the input and we would work it out from there?

Currently we have a `groupName` tag in all APIs as this is relied upon for the generators, I don't think we have a version tag (I remember looking up how they looked at the version and it falls back to package name), so we may want to enforce a `+groupVerison` tag or something, but yeah, will look into this